File: Graph Theory Peaks/Shortest Path/Q3. Design Graph With Shortest Path Calculator/Q3_DesignGraphWithShortestPathCalculator.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

class Graph {

    private List<List<int[]>> adjacencyList;
    // Adjacency list to store graph
    // Each node stores list of (neighbor, weight)

    public Graph(int n, int[][] edges) {

        adjacencyList = new ArrayList<>(n);
        // Create list to hold neighbors of each node

        for (int i = 0; i < n; i++) {
            adjacencyList.add(new ArrayList<>());
            // Initialize empty list for every node
        }

        for (int[] edge : edges) {
            addEdge(edge);
            // Add each edge to the graph
        }
    }

    public void addEdge(int[] edge) {

        int[] edgeArray = {edge[1], edge[2]};
        // edgeArray stores (destination, weight)

        adjacencyList.get(edge[0]).add(edgeArray);
        // Add edge from source to destination
    }

    public int shortestPath(int node1, int node2) {

        return dijkstra(node1, node2);
        // Call Dijkstra algorithm to find shortest path
    }

    private int dijkstra(int start, int end) {

        int n = adjacencyList.size();
        // Number of nodes

        int[] distances = new int[n];
        // distances[i] stores shortest distance from start to node i

        Arrays.fill(distances, Integer.MAX_VALUE);
        // Initialize all distances as infinity

        distances[start] = 0;
        // Distance to start node is 0

        // Priority queue to get node with smallest distance
        PriorityQueue<int[]> priorityQueue =
                new PriorityQueue<>(Comparator.comparingInt(o -> o[0]));

        priorityQueue.add(new int[]{0, start});
        // Add start node with cost 0

        while (!priorityQueue.isEmpty()) {

            int[] current = priorityQueue.poll();
            // Get node with smallest cost

            int currentNode = current[1], currentCost = current[0];
            // Extract node and its cost

            if (currentCost > distances[currentNode])
                continue;
            // Skip if a shorter path already exists

            if(currentNode == end)
                return currentCost;
            // If target reached return cost

            for (int[] edge : adjacencyList.get(currentNode)) {
                // Check all neighbors of current node

                int neighbor = edge[0], edgeLength = edge[1];
                // Neighbor node and edge weight

                int newRouteCost = edgeLength + distances[currentNode];
                // Total cost to reach neighbor

                if (distances[neighbor] > newRouteCost) {
                    // If shorter path found

                    distances[neighbor] = newRouteCost;
                    // Update shortest distance

                    priorityQueue.add(new int[]{newRouteCost, neighbor});
                    // Add neighbor to queue for further processing
                }
            }
        }

        return ((distances[end] == Integer.MAX_VALUE) ? -1 : distances[end]);
        // Return -1 if destination not reachable
    }
}

File: Graph Theory Peaks/Shortest Path/Q3. Design Graph With Shortest Path Calculator/Q3_DesignGraphWithShortestPathCalculator.py
class Graph(object):

    def __init__(self, n, edges):
        self.n = n  # number of nodes

        # create distance matrix and initialize with infinity
        self.distance = [[float('inf')] * n for _ in range(n)]

        # distance from node to itself is 0
        for i in range(n):
            self.distance[i][i] = 0

        # set direct edge distances
        for edge in edges:
            self.distance[edge[0]][edge[1]] = edge[2]

        # Floyd-Warshall algorithm to compute all-pairs shortest paths
        for k in range(n):          # intermediate node
            for i in range(n):      # start node
                for j in range(n):  # end node
                    self.distance[i][j] = min(
                        self.distance[i][j],                     # current shortest
                        self.distance[i][k] + self.distance[k][j]  # path through k
                    )

    def addEdge(self, edge):

        # if existing path is already shorter, ignore this edge
        if self.distance[edge[0]][edge[1]] <= edge[2]:
            return

        # update direct edge weight
        self.distance[edge[0]][edge[1]] = edge[2]

        # update shortest paths using the new edge
        for i in range(self.n):      # start node
            for j in range(self.n):  # end node
                self.distance[i][j] = min(
                    self.distance[i][j],  # current shortest path
                    self.distance[i][edge[0]] + edge[2] + self.distance[edge[1]][j]  # path using new edge
                )

    def shortestPath(self, node1, node2):

        # if no path exists
        if self.distance[node1][node2] == float('inf'):
            return -1

        # return shortest distance
        return self.distance[node1][node2]